keps/sig-node/3815-pod-condition-for-configuration-errors/README.md

```diff
@@ -0,0 +1,1021 @@
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+As a batch user, it is very likely that users experience configuration errors when submitting pods to Kubernetes.  We would like to propose a new condition so that users can determine if a pod has an error in startup.  A new condition will provide a quick way to verify if a pod detected a unrecoverable configuration error.
+
+## Motivation
+
+Batch users may not have deep knowledge of Kubernetes.  It is common occurrence to find users submitting Pods with configuration errors (invalid images, wrong image names, missing volumes, missing secrets, wrong volume/secret for config map).  These cases cause the Pod to be stuck in pending and there is code required to remove these pods or user input to remove.
+
+As a batch system developer, we find that higher level controllers have to write code to handle Pending Pod cases.  Since the pods are not considered failed, the higher level controller needs to transition this state to failed in order to enforce retries.  For example, the job controller will run these pods and the state of the Pod will be in Pending.  The job controller does not transition this state to failed even if the pod will never start running.
+
+For CNCF projects, in the Armada project, we implement logic that determines retry ability based on events and statuses.  Events are not standard so we expect issues when updating and we discovered various states that don't have a representative status.  
+
+It would be ideal to have a standard condition for these common configuration errors.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Add a new condition PodFailedToStartup to detect failures in configuration.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+- Will not modify behavior for ImagePullBackOff
```

Comment:
I don't know why we would exclude this?

Reply:
My thought for this is that there isn't enough information in pulling images to distinguish between the different failures:

1) If registry is overloaded I have seen cases where ImagePullBackOff occurs.  
2) If a docker image doesn't exist you get ImagePullBackOff.  
3) If you have invalid image pull secrets you also get ImagePullBackoff.  

There could be an interest in spelling these cases out more but it isn't as important case for us at the moment.  I wanted to focus on cases where the error requires human intervention to get it working again.  

In some of these cases I could see pulling images to work again if the registry is no longer overloaded.  But for 2-3 I guess these are cases that require human intervention but I don't know very much about how we pull images to distinguish between these different cases.

Reply:
We can tell the difference at SOME point, right?  Whether registry returns 503 or 404, for example.

But in either case, the pod is still "failing to start"

Reply:
yea I wasn't really going at it from a k8 dev perspective but from what is present in the pod yaml.  I'm not sure if the exit codes for the registry are exposed to a user.

Reply:
Threy are not, but we know them somewhere deep down